WordPress/gutenberg
# RichText: Fix empty value delete behaviors

Fixes #8731 

This pull request seeks to resolve two issues related to deletion behaviors on empty RichText (paragraph blocks). The first is described in #8731, the second I'm not sure has yet been surfaced (and exists in the production v3.4 release):

1. Navigate to Posts > Add New
2. Click writing prompt
3. Press Enter twice
4. Click first paragraph
5. Press (forward) Delete
6. Observe two paragraph deletion instead of one

The latter of these is due to the fact that we did not distinguish between `onMerge` and `onRemove` in the key handlers of RichText, thus causing _both_ to fire, merging into the one block and then deleting that one block, resulting in the double-deletion. This has been accounted for here by updating the block's `onMerge` to ~return a boolean indicating whether it was successful in its merge. This is interpreted by `RichText` to determine whether a removal is warranted, if possible~ only handle removal on the Backspace key (see https://github.com/WordPress/gutenberg/pull/8735#issuecomment-411776551).

The issue causing #8731 is a result of the fact that when focused in an empty RichText, the caret is placed adjacent (before) a "bogus" `br` node inserted by TinyMCE. The implementation of the general purpose `isHorizontalEdge` interprets this to mean that the caret is not at the end of the field, thus it would never allow the removal.

**Implementation notes:**

I adopted `waitForRichTextInitialization` from #8306 because I was having issues with failures locally without it. I'm not terribly happy with its inclusion, also because it only works for explicit inserts of the paragraph block, not a new paragraph inserted via Enter or other keyboard interactions.

**Testing instructions:**

Repeat testing instructions from #8731 and the above second set of issue steps, verifying that paragraph merges as expected, with a maximum of one removal.

Ensure end-to-end tests pass:

```
npm run test-e2e
```
